pydantic/_internal/_generate_schema.py

```diff
@@ -757,10 +757,14 @@ def set_discriminator(schema: CoreSchema) -> CoreSchema:
             'examples': field_info.examples,
         }
         json_schema_updates = {k: v for k, v in json_schema_updates.items() if v is not None}
-        json_schema_updates.update(field_info.json_schema_extra or {})
+        if isinstance(field_info.json_schema_extra, dict):
+            json_schema_updates.update(field_info.json_schema_extra)
```

Comment:
why don't we call `field_info.json_schema_extra` if it's a function here?

Reply:
okay so, in the update func, we have:
```python
            json_schema = {**handler(schema), **json_schema_updates}
```
and `handler(schema)` may result in a different schema than just the updates; `json_schema_updates` is not the whole json schema. In particular, getting stuff like `"type": "object"` (or even other types for `RootModel` subclasses), you probably want that to be accessible in the `json_schema_extra` call, which is why that needs to happen after the call to `handler` in the `json_schema_update_func` below.